File: backend/routers/metrics_router.py
"""
YUMMY Backend - Metrics Router
Endpoint: /metrics
"""

from fastapi import APIRouter
from config import DB

router = APIRouter(tags=["Metrics"])


@router.get("/metrics")
def get_metrics():
    """
    Xem request logs và chi phí AI.
    
    Hiển thị:
    - Tổng số request
    - Tổng chi phí USD (Gemini pricing)
    - 50 request gần nhất
    """
    logs = DB["request_logs"]
    total_cost = round(sum(l["cost"] for l in logs), 6)

    # Aggregate by agent
    agent_stats = {}
    for log in logs:
        agent = log["agent"]
        if agent not in agent_stats:
            agent_stats[agent] = {"calls": 0, "cost": 0.0, "total_tokens": 0}
        agent_stats[agent]["calls"] += 1
        agent_stats[agent]["cost"] += log["cost"]
        agent_stats[agent]["total_tokens"] += log.get("in_tokens", 0) + log.get("out_tokens", 0)

    return {
        "total_requests": len(logs),
        "total_cost_usd": total_cost,
        "agent_breakdown": agent_stats,
        "logs": logs[:50]  # 50 gần nhất
    }


@router.delete("/metrics")
def clear_metrics():
    """Xóa request logs."""
    DB["request_logs"] = []
    return {"status": "ok", "message": "Request logs đã được xóa."}

File: backend-ts/src/routers/metrics.router.ts
/**
 * Metrics router — /metrics.
 * Mirrors backend/routers/metrics_router.py.
 *
 * GET aggregates request_logs by agent and returns the 50 most recent rows.
 * DELETE clears all logs.
 */
import { createRoute, OpenAPIHono, z } from '@hono/zod-openapi';
import { type Bindings, createDb } from '../db/client.js';
import { logsRepo } from '../db/repositories/logs.repo.js';
import { MetricsResponseSchema, type RequestLogDto } from '../schemas/metrics.schema.js';

export const metricsRouter = new OpenAPIHono<{ Bindings: Bindings }>();

const json = <T extends z.ZodTypeAny>(schema: T) => ({
  'application/json': { schema },
});

interface AgentStats {
  calls: number;
  cost: number;
  total_tokens: number;
}

// ─── GET /metrics ────────────────────────────────────────
metricsRouter.openapi(
  createRoute({
    method: 'get',
    path: '/metrics',
    tags: ['Metrics'],
    responses: {
      200: { content: json(MetricsResponseSchema), description: 'Metrics' },
    },
  }),
  async (c) => {
    const db = createDb(c.env.DB);
    const logs = await logsRepo.list(db); // newest-first
    const totalCost = round(
      logs.reduce((sum, l) => sum + l.cost, 0),
      6,
    );

    const breakdown: Record<string, AgentStats> = {};
    for (const log of logs) {
      const a = log.agent;
      breakdown[a] ??= { calls: 0, cost: 0, total_tokens: 0 };
      const stats = breakdown[a];
      stats.calls += 1;
      stats.cost += log.cost;
      stats.total_tokens += log.inTokens + log.outTokens;
    }

    const dtoLogs: RequestLogDto[] = logs.slice(0, 50).map((l) => ({
      id: l.id,
      time: l.time,
      agent: l.agent,
      provider: l.provider,
      model: l.model,
      in_tokens: l.inTokens,
      out_tokens: l.outTokens,
      latency: l.latency,
      cost: l.cost,
    }));

    return c.json({
      total_requests: logs.length,
      total_cost_usd: totalCost,
      agent_breakdown: breakdown,
      logs: dtoLogs,
    });
  },
);

// ─── DELETE /metrics ─────────────────────────────────────
metricsRouter.openapi(
  createRoute({
    method: 'delete',
    path: '/metrics',
    tags: ['Metrics'],
    responses: {
      200: {
        content: json(z.object({ status: z.string(), message: z.string() })),
        description: 'Cleared',
      },
    },
  }),
  async (c) => {
    const db = createDb(c.env.DB);
    await logsRepo.clear(db);
    return c.json({ status: 'ok', message: 'Request logs cleared.' });
  },
);

function round(n: number, digits: number): number {
  const f = 10 ** digits;
  return Math.round(n * f) / f;
}
